protozfitsreader/tests/test_read_raw_events.py

```diff
@@ -0,0 +1,292 @@
+        event.trigger_output_patch19
+        event.baseline
+
+
+def test_event_number():
+    from protozfitsreader import ZFile
+
+    event_numbers = [
+        event.event_number
+        for event in ZFile(example_file_path)
+    ]
+    expected_event_numbers = [
+        FIRST_EVENT_IN_EXAMPLE_FILE + i
+        for i in range(EVENTS_IN_EXAMPLE_FILE)
+    ]
+    assert event_numbers == expected_event_numbers
+
+
+def test_telescope_ids():
+    from protozfitsreader import ZFile
+    telescope_ids = [
+        event.telescope_id
+        for event in ZFile(example_file_path)
+    ]
+    expected_ids = [TELESCOPE_ID_IN_EXAMPLE_FILE] * EVENTS_IN_EXAMPLE_FILE
+    assert telescope_ids == expected_ids
+
+
+def test_num_gains():
+    from protozfitsreader import ZFile
+    num_gains = [
+        event.num_gains
+        for event in ZFile(example_file_path)
+    ]
+    expected_num_gains = [0] * EVENTS_IN_EXAMPLE_FILE
+    assert num_gains == expected_num_gains
+
+
+def test_n_pixel():
+    from protozfitsreader import ZFile
+    n_pixel = [
+        event.n_pixels
+        for event in ZFile(example_file_path)
+    ]
+    assert n_pixel == [1296] * EVENTS_IN_EXAMPLE_FILE
+
+
+def test_pixel_flags():
+    from protozfitsreader import ZFile
+    pixel_flags = [
+        event.pixel_flags
+        for event in ZFile(example_file_path)
+    ]
+    expected_pixel_flags = [
+        np.ones(1296, dtype=np.bool)
+    ] * EVENTS_IN_EXAMPLE_FILE
+
+    for actual, expected in zip(pixel_flags, expected_pixel_flags):
+        assert (actual == expected).all()
+
+
+def test_local_time():
+    from protozfitsreader import ZFile
+    local_time = [
+        event.local_time
+        for event in ZFile(example_file_path)
+    ]
+    assert local_time == EXPECTED_LOCAL_TIME
+
+
+def test_gps_time():
+    from protozfitsreader import ZFile
+    gps_time = [
+        event.central_event_gps_time
+        for event in ZFile(example_file_path)
+    ]
+    assert gps_time == EXPECTED_GPS_TIME
+
+
+def test_camera_event_type():
+    from protozfitsreader import ZFile
+    camera_event_type = [
+        event.camera_event_type
+        for event in ZFile(example_file_path)
+    ]
+    assert camera_event_type == [1, 1, 1, 1, 1, 8, 1, 1, 1, 1]
+
+
+def test_array_event_type():
+    from protozfitsreader import ZFile
+    array_event_type = [
+        event.array_event_type
+        for event in ZFile(example_file_path)
+    ]
+    assert array_event_type == [0] * EVENTS_IN_EXAMPLE_FILE
+
+
+def test_num_samples():
+    from protozfitsreader import ZFile
+    num_samples = [
+        event.num_samples
+        for event in ZFile(example_file_path)
+    ]
+    assert num_samples == [50] * EVENTS_IN_EXAMPLE_FILE
+
+
+def test_adc_samples():
+    from protozfitsreader import ZFile
+    adc_samples = [
+        event.adc_samples
+        for event in ZFile(example_file_path)
+    ]
+
+    for actual in adc_samples:
+        assert actual.dtype == np.int16
```

Comment:
dtype should be uint16 not int16

Reply:
are you sure about this? Tests do not pass anymore, when I change this to uint16

Reply:
I am pretty sure that these are int16, from zfitsdump:
hiGain_waveforms_samples(int16[64800])

As far as i can remember, digicam indeed outputs unsigned ints up to 4096 (not sure why 4096 as afaik the ADC is 12 bits). So to prevent any big problem with conversions / early calculations I used signed integers to write the data. The idea was that if we end up with a negative number, then we know about it.

Reply:
There are 12bit ADCs with an overflow pin. In case of FACT the 12bit ADC can output all numbers between 0x000 and 0xFFF, but if the value is in the underflow range, the ADC will pull the overflow pin to HIGH. So **if** we write this overflow pin as an additional (most-significant) bit, the ADC can output:

 * in hex: 0x1000, 0x0000, 0x0001, ..., 0x0FFE, 0x0FFF, 0x1FFF
 * in dec: 4096, 0, 1, ..., 4094, 4095, 8191

of course "0x1000" is not "one more than 0x0FFF", but means: "underflow" and "0x1FFF" is not "4096 more than 0x0FFF" but means "overflow". 

could this be a similar behaviour here?

Reply:
I do not believe that we do have an overflow pin (Cyril knows better than me) because the PDP saturates first: if you look through very bright events, you will see a plateau in the pulse. Then in the anlysis we have a fit that reconstruct the real amplitude from the truncated pulse.

Reply:
Ok, then I have no idea how a 12-bit ADC can even output a number like 0x1000=4096

Reply:
>  I am pretty sure that these are int16, from zfitsdump: hiGain_waveforms_samples(int16[64800])

Ok then should we just convert it to uint16 while reading it (in `protozfitsreader`) ? (reducing memory) or we just keep it that way ? I think so far in `digicampipe` I did, in some places, the int to uint transformation. 
Also since usually we set ~0 p.e. signal to 200 ADC (1 p.e. ~ 5.8 ADC) that way we avoid negative values (underflow).
Concerning overflow : yes the PDP (pre-amp) saturates before reaching the FADC limit : 
pre-amp saturation ~ 400 p.e. = 200 + 5.8*400 = 2520 ADC < 4095 ADC.

 > Then in the anlysis we have a fit that reconstruct the real amplitude from the truncated pulse.

Yes

Reply:
I thought python would do an "auto-magical" transformation from uint to it if one does : 

```bash
>>> import numpy as np
>>> a = np.array([0, 2, 3, 6], dtype=np.uint16)
>>> a
array([0, 2, 3, 6], dtype=uint16)
>>> a - 3
array([65533, 65535,     0,     3], dtype=uint16)
>>> a = a - 3
>>> a
array([65533, 65535,     0,     3], dtype=uint16)
>>> 
```

indeed no magic here. So it can cause problems in the pipeline if we don't assign the correct type for each field of the containers...

Reply:
You code the pipeline, so eventually it will be up to you. Myself I would keep everything as signed integers because we are far from using the whole range of signed ints, and being able to move to negative numbers is always a plus from my experience: uint is really useful only when needed to extend the range of values. 
Yes, "conservatives" coders might argue that the values coming in really are uints, and that they should be treated as such. I disagree, most probably because I had to struggle against "negative values with uints" more often than I like to admit.

Reply:
>>>  I disagree, most probably because I had to struggle against "negative values with uints" more often than I like to admit.
 
😃 

Ok let's keep it that way but we have to look how `ctapipe`will handle it.

Reply:
I think signed ints are very useful here. At some point, the "baseline" is subtracted from the adc samples and then the result is very likely to contain negative numbers. 

When both players are signed 16bit integers subtracting them is fine and fast:
```python
In [1]: np.array([1], dtype='i2') - np.array([5], dtype='i2')
Out[1]: array([-4], dtype=int16)
```

but when one is unsigned 16bit, the result has to be signed 32bit.
```python
In [2]: np.array([1], dtype='i2') - np.array([5], dtype='u2')
Out[2]: array([-4], dtype=int32)
```

Of course we can still suffer from overflows:
```
In [3]: np.array([-3], dtype='i2') - np.array([-2**15], dtype='i2')
Out[3]: array([32765], dtype=int16)
```

Reply:
> but when one is unsigned 16bit, the result has to be signed 32bit.

good point

Reply:
We must just keep in mind, that adc value = -1 must not be used to mark something like "integer nan" ... since it may occur as a normal result of some calculation.

Reply:
can be ignored, c.f. comment below